### PR TITLE
2620 item link on sync simplification

### LINKED ISSUE #2620: Item merging: simplify item_link generation in sync
## Is your feature request related to a problem? Please describe 👀

@andreievg has suggested that `item_link` generation on sync can simplified by just doing as part of the item repository rather than adding a `ItemLinkTranslation` and file. Sounds neat!

## Describe the solution you'd like 🎁

Remove [item_link.rs](https://github.com/msupply-foundation/open-msupply/blob/2771e7c907108bc376028460d17d48d6edce3d78/server/service/src/sync/translations/special/item_link.rs) and modify [item_row.rs](https://github.com/msupply-foundation/open-msupply/blob/2771e7c907108bc376028460d17d48d6edce3d78/server/repository/src/db_diesel/item_row.rs) to insert an item link when upserting or inserting. If conflict for the item_link because it already exists, just do nothing.

### Moneyworks Jobcode 🧰

OMS:MERGE

### PR DESCRIPTION
**Pull request by Chris-Petty**
Fixes #2620 

# 👩🏻‍💻 What does this PR do? 

Removed the `item_link.rs` special translator and moved `item_link` creation to be in the `ItemRowRepo`. Less code and at a lower level as to avoid accidentally skirting around it (e.g. a future mSupply central server could likely rely on this new implementation)

# 🧪 How has/should this change been tested? 

Just running tests so far.

### CODE REVIEW

File: server/repository/src/db_diesel/item_row.rs [hunk 1]
```diff
@@ -85,6 +97,8 @@ impl<'a> ItemRowRepository<'a> {
             .do_update()
             .set(item_row)
             .execute(&self.connection.connection)?;
+
+        insert_or_ignore_item_link(&self.connection, item_row)?;
```
Comment (Chris-Petty) [inline server/repository/src/db_diesel/item_row.rs:101]: We could possibly remove `insert_or_ignore_item_link` call a one-liner instead pretty cleanly if we implemented  `ItemLinkRow::from(ItemRow)` trait. I did go down this road but I didn't know what the expectation is around when to use references:

```rs
// item_link_row.rs
impl From<ItemRow> for ItemLinkRow {
    fn from(item: ItemRow) -> Self {
        Self {
            id: item.id.clone(),
            item_id: item.id.clone(),
        }
    }
}

// Here
ItemLinkRowRepository::new(&self.connection)
            .insert_one_or_ignore(&ItemLinkRow::from(item_row.clone()))?;
```
Sooo much cloning though? Kinda ugly? OR:

```rs
// item_link_row.rs
impl From<&ItemRow> for ItemLinkRow {
    fn from(item: &ItemRow) -> Self {
        Self {
            id: item.id.clone(),
            item_id: item.id.clone(),
        }
    }
}

// Here
ItemLinkRowRepository::new(&self.connection)
            .insert_one_or_ignore(&ItemLinkRow::from(item_row))?;
```

So naturally passing a reference has less cloning, but is it normal to implement `From<&T>` or `From<T>`?

File: server/repository/src/db_diesel/item_link_row.rs [hunk 2]
```diff
@@ -64,6 +64,24 @@ impl<'a> ItemLinkRowRepository<'a> {
         Ok(())
     }
 
+    #[cfg(feature = "postgres")]
+    pub fn insert_one_or_ignore(&self, item_link_row: &ItemLinkRow) -> Result<(), RepositoryError> {
+        diesel::insert_into(item_link_dsl::item_link)
+            .values(item_link_row)
+            .on_conflict(item_link::id)
```
Comment (Chris-Petty) [inline server/repository/src/db_diesel/item_link_row.rs:71]: Be kinda nice if diesel just had an `insert_one_or_ignore` implementation for pgsql that was just like this 🤔... looks like they've gone the other way and settled on the `on_conflict` method for sqlite and pgsql in [2.1.4](https://docs.rs/diesel/2.1.4/diesel/query_builder/struct.InsertStatement.html#method.on_conflict)

File: server/repository/src/db_diesel/item_row.rs (hunk 1 above)
Comment (Chris-Petty) [reply server/repository/src/db_diesel/item_row.rs:101]: I suppose it wouldn't be incorrect to do both!

Comment (andreievg) [reply server/repository/src/db_diesel/item_row.rs:101]: I think cloning should be fine here, and it's pretty clear ot me, without clone that there is a conversion from item row to item link row. I think from becomes usefull for common things where logic is repeated many times over, or maybe for generic things that do something to unknown data type that has to have a 'from' property (like collect() for iterator)

**Review by andreievg — APPROVED** [review]
Looks good to me, thanks Chris

**Pull request merged**